keymanager/src/client/remote.rs

```diff
@@ -584,4 +596,102 @@ impl KeyManagerClient for RemoteClient {
             .await
             .map_err(|err| KeyManagerError::Other(err.into()))
     }
+
+    async fn state_key(
+        &self,
+        churp_id: u8,
+        key_id: KeyPairId,
+    ) -> Result<StateKey, KeyManagerError> {
+        let id = (key_id, churp_id);
+
+        // First try to fetch from cache.
+        {
+            let mut cache = self.state_keys.write().unwrap();
+            if let Some(key) = cache.get(&id) {
+                return Ok(key.clone());
+            }
+        }
+
+        // No entry in cache, fetch from key manager.
+        let consensus_state = self.consensus_verifier.latest_state().await?;
+        let registry_state = consensus_state.registry_state();
+        let key_manager_state = consensus_state.key_manager_state();
+        let churp_state = key_manager_state.churp();
+
+        let status = tokio::task::block_in_place(move || {
+            let runtime_id = registry_state.key_manager(&self.runtime_id)?;
+            churp_state.status(runtime_id, churp_id)
+        })?;
+
+        let mut committee = status.committee;
+        committee.shuffle(&mut OsRng);
+
+        let kind = HandoffKind::CommitteeUnchanged;
+        let player = Player::new(status.threshold, kind);
+        let min_shares = player.min_shares();
+        let mut encoded_shares = Vec::with_capacity(min_shares);
+
+        // TODO: Optimize by fetching key shares concurrently.
+        for node_id in committee {
+            if encoded_shares.len() == min_shares {
+                break;
+            }
+
+            self.rpc_client
+                .update_nodes_async(vec![node_id])
+                .await
+                .map_err(|err| KeyManagerError::Other(err.into()))?;
+
+            let encoded_share: EncodedEncryptedPoint = self
+                .rpc_client
+                .secure_call(
+                    METHOD_KEY_SHARE,
+                    KeyShareRequest {
+                        id: status.id,
+                        runtime_id: status.runtime_id,
+                        epoch: status.handoff,
+                        key_id,
+                    },
+                )
+                .await
+                .into_result_with_feedback()
+                .await
+                .map_err(|err| KeyManagerError::Other(err.into()))?;
+
+            encoded_shares.push(encoded_share);
+        }
+
+        let mut salt = [0; NAMESPACE_SIZE + 1 + KEY_PAIR_ID_SIZE];
+        salt[..NAMESPACE_SIZE].copy_from_slice(&status.runtime_id.0);
+        salt[NAMESPACE_SIZE] = status.id;
+        salt[NAMESPACE_SIZE + 1..].copy_from_slice(&key_id.0);
+
+        let state_key = match status.suite_id {
+            SuiteId::NistP384Sha3_384 => {
+                let mut shares = Vec::with_capacity(min_shares);
+                for encoded_share in encoded_shares {
+                    let share = encoded_share
+                        .try_into()
+                        .map_err(|err: Error| KeyManagerError::Other(err.into()))?;
```

Comment:
This should probably done while fetching shares to make sure enough shares are fetched in case some cannot be decoded?